src/App.jsx

```diff
@@ -1,37 +1,245 @@
+
+      const { records } = await resp.json();
+      const savedTodo = {
+        id: records[0].id,
+        ...records[0].fields,
+      };
+
+      if (!records[0].fields.isCompleted) {
+        savedTodo.isCompleted = false;
+      }
+
+      setTodoList([...todoList, savedTodo]);
+    } catch (error) {
+      console.log(error);
+      setErrorMessage(error.message);
+    } finally {
+      setIsSaving(false);
+    }
+  };
+
+  const completeTodo = async (id) => {
+    const inProgressTodo = todoList.find((todo) => todo.id === id);
+
+    const payload = {
+      records: [
+        {
+          id: inProgressTodo.id,
+          fields: {
+            title: inProgressTodo.title,
+            isCompleted: true,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error('Error completing');
+      }
+
+      const { records } = await resp.json();
+
+      const completedTodo = {
+        id: records[0]['id'],
+        ...records[0].fields,
+      };
+
+      if (!records[0].fields.isCompleted) {
+        completedTodo.isCompleted = false;
+      }
+
+      const completedTodos = todoList.map((todo) =>
+        todo.id === completedTodo.id ? { ...completedTodo } : todo
+      );
+
+      setTodoList(completedTodos);
+    } catch (error) {
+      console.log(error);
+
+      setErrorMessage(`${error.message}. Reverting todo...`);
+
+      const revertedTodos = todoList.map((todo) =>
+        todo.id === id ? { ...todo, isCompleted: false } : todo
+      );
+
+      setTodoList([...revertedTodos]);
+    } finally {
+      setIsSaving(false);
+    }
+  };
+
+  const updateTodo = async (editedTodo) => {
+    const originalTodo = todoList.find((todo) => todo.id === editedTodo.id);
+
+    const payload = {
+      records: [
+        {
+          id: editedTodo.id,
+          fields: {
+            title: editedTodo.title,
+            isCompleted: editedTodo.isCompleted,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error('Error updating');
+      }
+
+      const { records } = await resp.json();
+      const updatedTodo = {
+        id: records[0]['id'],
+        ...records[0].fields,
+      };
+
+      if (!records[0].fields.isCompleted) {
+        updatedTodo.isCompleted = false;
+      }
+
+      const updatedTodos = todoList.map((todo) =>
+        todo.id === updatedTodo.id ? { ...updatedTodo } : todo
+      );
+
+      setTodoList([...updatedTodos]);
```

Comment:
This is a small bug. There is noticeable lag when a todo is edited or completed. Consider when the state is being set to solve this.